### ISSUE TITLE
some questions

### ISSUE THREAD

**Issue by shams2023** [zer0int/CLIP-text-image-interpretability#1]
Hello author
May I ask which article did you mention this visualization operation in? I really need him, thank you!

**Comment by zer0int** [comment]
Hi @shams2023 - I am not sure what you are referring to. I only shared the link on social media, but I never wrote an "article" / e.g. blog post about it. Maybe somebody else did - can you give me some more details about what you are looking for? Do you need any specific information? Kind regards!

**Comment by shams2023** [comment]
> Hi @shams2023 - I am not sure what you are referring to. I only shared the link on social media, but I never wrote an "article" / e.g. blog post about it. Maybe somebody else did - can you give me some more details about what you are looking for? Do you need any specific information? Kind regards!

Hello author!
I just want to know, what task is the main purpose of the code you uploaded?
Can the following tasks be completed:
Can we visualize which part of the image the text focuses on by using the clip model to obtain the corresponding modal feature representation of the image text pair?
Thank you for your reply. Wishing you success in your work and studies!
（你好作者！
我其实就是想知道，你上传的代码，主要是为了完成什么任务？
能否完成如下工作：
图像-文本对通过clip模型得到对应模态的特征表示，能否可视化文本到底关注到了图像的哪一部分吗？
感谢你的回复，祝你工作学业顺利！）

**Comment by zer0int** [comment]
@shams2023 
With attention visualization (alike "GradCAM"), you would normally provide the class, such as "cat" or "a photo of a dog" as a text input, together with the image input. Then, you could get an attention heat map for the features CLIP identifies in the image as most relevant to the class. So you could find that CLIP's attention is mostly on the eyes for "cat", but might be focused on a dog's ear as most relevant feature for 'dog'.

However, this basically "forces" CLIP to see a pre-defined class (text provided by human user).

This code, however, first uses CLIP gradient ascent on the input image -> output is text tokens / words, the prediction of a class / classes that CLIP itself "sees" in the image. It basically obtains CLIP's "opinion" about what the image contains, according to the features CLIP has learned during training. For a photo of a cat, CLIP may also predicts "cat" as one the classes - but CLIP predicts further words / classes, some that might seem unusual or non-sensical to humans, such as "domino" or "stripe" or "fanci" or "catmented", as CLIP will identify local patterns (it might see a "map" in a cat fur pattern, too), as well as background patterns that humans might not be paying attention to:

![F9naLNhWAAABTV_](https://github.com/zer0int/CLIP-text-image-interpretability/assets/132047210/d836f2aa-6b6e-4ad4-a1ae-a2cb7d0c5b18)

It also reveals what is known as CLIP's typographic attack vulnerability, the general tendency of the model to overfit on text (but it recognized both "apple" and "ipod" in this famous example image from OpenAI's blog - as well as "fakepods"). This will result in a very different attention heatmap, that shows very well how CLIP is overly focused on text - especially when you compare it to the attention when the text has been removed (photoshopped):

![F9nlAYdX0AAXiVK](https://github.com/zer0int/CLIP-text-image-interpretability/assets/132047210/cd1308df-3106-4843-9bcc-b911793731d1)


So, to summarize:
1. We get CLIP's own "text opinion" about what CLIP mainly "sees" in the image (features and associated tokens / classes), and don't tell CLIP what it must see in the image.
2. Then we use CLIP's own predicted words / "opinion" about the salient features in the image to create the heatmaps of [text-class]-[image] pairs.

The main difference is that we do not tell CLIP what it should see in the image (by defining the text / class it should "look for"), but we get CLIP's own "opinion" (text prediction) of what the image contains first - and then we visualize that as an attention heatmap.

I hope this helps! 👋

**Comment by shams2023** [comment]
> 使用注意力可视化（类似于“GradCAM”），您通常会将类（例如“猫”或“狗的照片”）作为文本输入，以及图像输入。然后，您可以获得 CLIP 在图像中标识为与类最相关的特征的注意力热图。因此，您会发现 CLIP 的注意力主要集中在“猫”的眼睛上，但可能将注意力集中在狗的耳朵上，因为与“狗”最相关的特征。
> 
> 但是，这基本上“强制”CLIP看到预定义的类（由人类用户提供的文本）。
> 
> 然而，这段代码首先在输入图像上使用 CLIP 梯度上升，>输出是文本标记/单词，即 CLIP 本身在图像中“看到”的类/类的预测。它基本上根据 CLIP 在训练中学到的特征来获得 CLIP 对图像包含的内容的“意见”。对于猫的照片，CLIP 也可能将“猫”预测为类别之一 - 但 CLIP 预测其他单词/类别，一些对人类来说可能看起来不寻常或无意义，例如“多米诺骨牌”或“条纹”或“幻想”或“猫”，因为 CLIP 会识别局部模式（它也可能在猫毛图案中看到“地图”）， 以及人类可能没有注意的背景模式：
> 
> ![F9naLNhWAAABTV_](https://user-images.githubusercontent.com/132047210/284224464-d836f2aa-6b6e-4ad4-a1ae-a2cb7d0c5b18.png)
> 
> 它还揭示了所谓的 CLIP 的排版攻击漏洞，即该模型在文本上过度拟合的一般趋势（但它在 OpenAI 博客的这个著名示例图像中识别了“apple”和“ipod”——以及“fakepods”）。这将导致一个非常不同的注意力热图，它很好地显示了 CLIP 是如何过度关注文本的——尤其是当你将其与文本被删除时的注意力进行比较（photoshopped）：
> 
> ![F9nlAYdX0AAXiVK](https://user-images.githubusercontent.com/132047210/284225888-cd1308df-3106-4843-9bcc-b911793731d1.jpg)
> 
> 所以，总结一下：
> 
> 1. 我们得到 CLIP 自己的“文本意见”，即 CLIP 在图像中主要“看到”什么（特征和相关标记/类），而不是告诉 CLIP 它必须在图像中看到什么。
> 2. 然后，我们使用 CLIP 自己对图像中显着特征的预测词/“意见”来创建 [text-class]-[image] 对的热图。
> 
> 主要区别在于，我们没有告诉 CLIP 它应该在图像中看到什么（通过定义它应该“寻找”的文本/类），但我们首先得到 CLIP 自己对图像包含的内容的“意见”（文本预测）——然后我们将其可视化为注意力热图。
> 
> 我希望这会有所帮助！👋

https://github.com/zer0int/CLIP-text-image-interpretability/issues/1#issuecomment-1818727562

**Comment by shams2023** [comment]
> With attention visualization (similar to "GradCAM"), you'll typically have a class (such as a "cat" or "dog photo") as text input, as well as an image input. You can then get an attention heatmap of the features that CLIP identifies in the image as the most relevant to the class. As a result, you'll find that CLIP's attention is primarily focused on the "cat's" eyes, but may focus on the dog's ears because of the features most associated with the "dog."
> 
> However, this basically "forces" CLIP to see a predefined class (text provided by a human user).
> 
> However, this code first uses a CLIP gradient ascent on the input image, and the > output is a text marker/word, i.e. a prediction of the class/class that the CLIP itself "sees" in the image. It basically gets CLIP's "opinion" of what the image contains based on the features that CLIP learns in training. For photos of cats, CLIP may also predict "cat" as one of the categories - but CLIP predicts other words/categories, some that may seem unusual or meaningless to humans, such as "dominoes" or "stripes" or "fantasies" or "cats" because CLIP recognizes local patterns (it may also see "maps" in cat hair patterns), as well as background patterns that humans may not be paying attention to:
> 
> ![F9naLNhWAAABTV_](https://user-images.githubusercontent.com/132047210/284224464-d836f2aa-6b6e-4ad4-a1ae-a2cb7d0c5b18.png)
> 
> It also reveals the so-called typographic exploit vulnerability of CLIP, the general tendency of the model to overfit textually (but it identifies "apple" and "ipod" – as well as "fakepods" – in this famous example image from the OpenAI blog). This will result in a very different attention heatmap, which does a good job of showing how CLIP is overly focused on text – especially when you compare it to the attention when the text is deleted (photoshopped):
> 
> ![F9nlAYdX0AAXiVK](https://user-images.githubusercontent.com/132047210/284225888-cd1308df-3106-4843-9bcc-b911793731d1.jpg)
> 
> So, to sum it up:
> 
> 1. We get CLIP's own "textual opinion", i.e. what the CLIP primarily "sees" in the image (features and associated tags/classes), rather than telling the CLIP what it must see in the image.
> 2. We then use CLIP's own prediction words/"opinions" of salient features in the image to create a heatmap of the [text-class]-[image] pairs.
> 
> The main difference is that we don't tell CLIP what it should see in the image (by defining the text/class it should be "looking for"), but we first get CLIP's own "opinion" (text prediction) of what the image contains – and then we visualize it as an attention heat map.
> 
> I hope this helps! 👋

Thank you very much for your help. Thank you again, thank you!
